### PR TITLE
Feature/decoder only as formatter

### PR DESCRIPTION
**Pull request by elboy3**
> **Please do not create a pull request without creating an issue first.**
>
> Changes need to be discussed before proceeding, pull requests submitted without linked issues may be rejected.
>
> Please provide enough information so that others can review your pull request. You can skip this if you're fixing a typo – it happens.

* [ ] I have added tests to `tests` to cover my changes.
* [ ] I have updated `docs/`, if necessary.
* [ ] I have updated the `README.md`, if necessary.

***What existing issue does this pull request close?***

Put `closes #issue-number` in this pull request's description to auto-close the issue that this fixes.

***How are these changes tested?***

This pull request includes automated tests for the code it touches and those tests are described below. If no tests are included, reasons why must be provided below.

These changes are tested with [...]

***Demonstration***

Demonstrate your contribution.

For example, what are the exact commands you ran and their output, related screenshots, screen-recordings, test runs, anything that can showcase.

***Provide additional context.***

Provide as much relevant context as you like.

### CODE REVIEW

File: dataquality/integrations/seq2seq/hf.py [hunk 1]
```diff
@@ -89,6 +115,16 @@ def watch(
 
     set_tokenizer(tokenizer, max_input_tokens, max_target_tokens)
 
+    # Set the response template for DecoderOnly models
+    if response_template:
```
Comment (bogdan-galileo) [inline dataquality/integrations/seq2seq/hf.py:119]: don't we need to warn them if the model is decoder only and they forget to pass response_template ?

File: dataquality/schemas/task_type.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ class TaskType(str, Enum):
     object_detection = "object_detection"
     semantic_segmentation = "semantic_segmentation"
     prompt_evaluation = "prompt_evaluation"
-    seq2seq = "seq2seq"
+    seq2seq = "seq2seq"  # deprecated, use encoder_decoder or decoder_only
```
Comment (bogdan-galileo) [inline dataquality/schemas/task_type.py:17]: what do you mean ? Where are these new tasks ?

File: dataquality/utils/seq2seq/__init__.py [hunk 3]
```diff
@@ -15,16 +17,15 @@ def remove_padding(
...
-        non-padding logits.
-    padding_side: str
-        Comes from the tokenizer used for the model, determines
-        which side padding is applied.
     padded_token_seq: np.ndarray of shape - [max_seq_len, ...]
         Padded token sequence. The first dimension must be the token
         dimension and be >= num_tokens. The following dimensions are
         unrestricted.
+    num_tokens: int
+        Length of the non-padding logits.
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq/__init__.py:25]: -> non-padded

File: dataquality/loggers/model_logger/base_model_logger.py [hunk 4]
```diff
@@ -111,7 +110,9 @@ def log(self) -> None:
         # global variables (cur_split and cur_epoch) that are subject to change
         # between subsequent threads
         self.set_split_epoch()
-        ThreadPoolManager.add_thread(target=self._add_threaded_log)
+        # TODO Undo
```
Comment (bogdan-galileo) [inline dataquality/loggers/model_logger/base_model_logger.py:113]: UNDO ?

File: dataquality/loggers/logger_config/seq2seq/seq2seq_base.py [hunk 5]
```diff
@@ -17,6 +19,24 @@ class Seq2SeqLoggerConfig(BaseLoggerConfig):
...
+    model_type: Seq2SeqModelTypes = Seq2SeqModelTypes.encoder_decoder
+    # Decoder only below
+    id_to_formatted_prompt_length: Dict[str, Dict[int, int]] = defaultdict(dict)
+    response_template: Optional[Union[str, List[int]]] = None
+
+    @validator("response_template")
+    def ensure_tokenized_response_template(
+        cls, response_template: Optional[Union[str, List[int]]], values: Dict[str, Any]
+    ) -> List[int]:
+        if response_template and isinstance(response_template, str):
```
Comment (bogdan-galileo) [inline dataquality/loggers/logger_config/seq2seq/seq2seq_base.py:31]: do we need an elif for when it's a list of int ?

File: dataquality/utils/seq2seq/decoder_only.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    Example:
+          >> formatted_response = [7, 1, 2, 3, 8, 5, 9, 1, 2, 3, 9, 10, 6]
+          >> response_template = [1, 2, 3]
+          >> extract_tokenized_responses(formatted_response, response_template)
+            ...
+            [9, 10, 6]
+            ...
+
+    If a sample does not contain the response_template we represent the
+    tokenized_response for that sample as [] - i.e. the <Empty String>.
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq/decoder_only.py:28]: Random: do we need something like an end of sequence, or anything like this that all the non-empty targets have ?

File: dataquality/utils/seq2seq/decoder_only.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    Parameters:
+        TODO
+    """
+    tokenized_responses: List[List[int]] = []
+    for t_prompt in tokenized_formatted_prompts:
+        # Reverse search over matches of the first token in the response template
+        matched_indices = np.where(np.array(t_prompt) == response_template[0])[0]
+        response_token_ids_start_idx = None
+        for i in range(len(matched_indices)):
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq/decoder_only.py:38]: We should traverse that one in reverse and if we find, just exit from the for loop (the first character could be common and `matched_indices` could be large!)

File: dataquality/utils/seq2seq/decoder_only.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    tokenized_response for that sample as [] - i.e. the <Empty String>.
+    """
+    tokenized_responses: List[List[int]] = []
+    for t_prompt in tokenized_formatted_prompts:
+        if not response_template:
+            tokenized_responses.append([])
+            continue
+
+        # Reverse search over matches of the first token in the response template
+        matched_indices = np.where(np.array(t_prompt) == response_template[0])[0]
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq/decoder_only.py:35]: Testing with 50K rows of length 512 and vocab = 30K, this np.where takes almost 2sec.

We can do it instant with 1 numpy call returning a 2d matrix of indices found

Screenshot 2023-11-13 at 1 44 06 PM

File: dataquality/utils/seq2seq/decoder_only.py (hunk 7 above)
Comment (bogdan-galileo) [reply dataquality/utils/seq2seq/decoder_only.py:37]: we could actually also combine the below code too. If we found that match, do what's needed, else warn that we didn't find

File: dataquality/loggers/model_logger/seq2seq/seq2seq_base.py [hunk 9]
```diff
@@ -110,33 +120,18 @@ def process_logprobs(
...
-        ):
-            sample_n_tokens = sample_logprobs.shape[0]
-            # Truncating the tokens computed in dq to ensure that there aren't more than
-            # what the model outputed (or we get an indexing error).
-            sample_labels = self._retrieve_sample_labels(
-                sample_id, max_tokens=sample_n_tokens
-            )
-            padding_side = self.logger_config.tokenizer.padding_side
-            sample_logprobs = remove_padding(
+        for sample_id, sample_logprobs in zip(batch_ids, batch_logits):
```
Comment (bogdan-galileo) [inline dataquality/loggers/model_logger/seq2seq/seq2seq_base.py:129]: @jonathangomesselman are these `sample_logprobs` or `sample_logits`, or sometimes one and sometimes the other ?

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py [hunk 10]
```diff
@@ -163,6 +178,7 @@ def log_dataset(
         text: Union[str, int] = "text",
         id: Union[str, int] = "id",
         label: Optional[Union[str, int]] = "label",
+        formatted_prompt: Union[str, int] = "formatted_label",
```
Comment (bogdan-galileo) [inline dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:182]: Why not default to `galileo_formatted_prompt` ?

File: dataquality/integrations/seq2seq/hf.py (hunk 1 above)
Comment (jonathangomesselman) [reply dataquality/integrations/seq2seq/hf.py:119]: Yeah good call! We might want to actually just fail / throw an exception

File: dataquality/loggers/logger_config/seq2seq/seq2seq_base.py (hunk 5 above)
Comment (jonathangomesselman) [reply dataquality/loggers/logger_config/seq2seq/seq2seq_base.py]: If it is already a list of ints then we are all good?

File: dataquality/utils/seq2seq/decoder_only.py (hunk 6 above)
Comment (jonathangomesselman) [reply dataquality/utils/seq2seq/decoder_only.py:26]: I thought about this! The thing is that if the response template isn't found then their truly are no tokens in the `target`. Like the model has run out of tokens so I think that this is technically what we have to do? I agree I was a bit unsure for this edge case.

File: dataquality/utils/seq2seq/decoder_only.py (hunk 7 above)
Comment (jonathangomesselman) [reply dataquality/utils/seq2seq/decoder_only.py:37]: We are going in reverse order! But yes I did mean to break out of this

File: dataquality/loggers/model_logger/seq2seq/seq2seq_base.py (hunk 9 above)
Comment (jonathangomesselman) [reply dataquality/loggers/model_logger/seq2seq/seq2seq_base.py:129]: Good call! This should now be logits

Comment (jonathangomesselman) [reply dataquality/loggers/model_logger/seq2seq/seq2seq_base.py:129]: Speaking of. I need to pull down this branch and add some local changes I have been making in a separate branch.

File: dataquality/utils/seq2seq/decoder_only.py (hunk 8 above)
Comment (bogdan-galileo) [reply dataquality/utils/seq2seq/decoder_only.py:35]: awesome, doing now

File: dataquality/loggers/data_logger/seq2seq/seq2seq_base.py (hunk 10 above)
Comment (jonathangomesselman) [reply dataquality/loggers/data_logger/seq2seq/seq2seq_base.py:182]: I think `galileo_formatted_prompt` is more for internal use?

File: dataquality/utils/seq2seq/decoder_only.py (hunk 8 above)
Comment (bogdan-galileo) [reply dataquality/utils/seq2seq/decoder_only.py:35]: btw that comparison was not fair, np.where is not doing what I though. Still we can make that faster with argwhere for ex

File: dataquality/integrations/seq2seq/hf.py (hunk 1 above)
Comment (jonathangomesselman) [reply dataquality/integrations/seq2seq/hf.py:120]: Added this!

File: dataquality/loggers/model_logger/seq2seq/seq2seq_base.py [hunk 11]
```diff
@@ -46,18 +57,22 @@ def __init__(
             epoch=epoch,
             inference_name=inference_name,
         )
+        self.logprobs = logprobs if logprobs is not None else []
```
Comment (jonathangomesselman) [inline dataquality/loggers/model_logger/seq2seq/seq2seq_base.py:60]: Note this will be getting cleaned up as I officially add support for logging `logprobs`

File: dataquality/utils/seq2seq/decoder_only.py (hunk 6 above)
Comment (bogdan-galileo) [reply dataquality/utils/seq2seq/decoder_only.py:26]: sounds good!

File: dataquality/utils/seq2seq/decoder_only.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                    response_token_ids_start_idx + len(response_template) :
+                ]
+                break
+
+        tokenized_responses.append(tokenized_response)
+
+        # Warn if the response template was not found!
+        if response_token_ids_start_idx is None:
+            warn(
+                f"Could not find response key `{response_template}` in the "
```
Comment (bogdan-galileo) [inline dataquality/utils/seq2seq/decoder_only.py:56]: @jonathangomesselman last point: since response_template and t_prompt are just lists of token_ids, do you think it's worth printing them ?

File: dataquality/loggers/model_logger/seq2seq/formatters.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+        # Restrict to just the response tokens
+        num_response_tokens = len(response_labels)
+        # TODO check - Shift the logits such that tokens < n predict token n.
+        #   notice here that we ignore the final token logprob since there is
+        #   no n+1 token. For DecoderOnly the logits and labels are implicitly
+        #   shifted within the model.
+        sample_logits = sample_logits[-(num_response_tokens + 1) : -1]
+
+        sample_logprobs = self.convert_logits_to_logprobs(sample_logits)
```
Comment (jonathangomesselman) [inline dataquality/loggers/model_logger/seq2seq/formatters.py:172]: @bogdan-galileo @elboy3 In the future I am planning to move this outside of this function. This makes the function purely for formatting - i.e. removing padding and restricting to response tokens. This makes it a bit cleaner and also allows us to share logic when the user logs just logprobs!

**Comment by elboy3** [comment]
Closing for @jonathangomesselman to open a new PR !
